### ISSUE TITLE
FIDO Payment Support

### ISSUE THREAD

**Issue by cyberphone** [w3c/webauthn#1570]
Since the SPC specification is to be discussed at https://github.com/w3c/webpayments/wiki/Agenda-FTF2021, I took the liberty outlining an enhanced scheme that may serve as "inspiration".  It is not entirely complete (enrollment is currently missing), but the core is hopefully understandable.  The proposal mainly deals with security constructs and data flows. 

Updated: 2021-02-19

### Enhancements

- _Privacy_: User-unique information (PII) is neither required nor provided to external actors.
- _Security_: Reduced attack space since no sensitive data is given to external actors. Encryption replaces tokenization.
- _Universality_: Applicable to any account based payment system.
- _Simplicity_:  In similarity to EMV the scheme does not not depend on external challenge data or centralized registries holding enrolled account numbers.

Although not a part of this specification, the underpinning authorization scheme also opens the door to non-direct payments, including secure (tokenized) versions of Card-on-File. 

### Prerequisite - Payment Credentials
The scheme outlined here depends on that the **User** already has received a payment credential from the **Issuer** in an _enrollment process_.   A payment credential is supposed to at least contain the following elements:
- A **User** specific (unique) FIDO key for _signing_ payment authorization data.
- An `accountId` associated with the FIDO signature key.
- An **Issuer** specific (shared) public key for _encrypting_ payment authorization data.
- An `issuerId` in the form of a BIN number or a URL associated with the **Issuer**.
- An icon identifying the payment credential for the **User**.

A payment credential is associated with a specific `paymentMethod` established during enrollment.

### Detailed Operation
The different steps in following state diagram are described in this section.
![state-diagram](https://user-images.githubusercontent.com/8044211/107975497-7dca2f00-6fb8-11eb-9e3d-08d1fff16ddf.png)
Note: this specification does not elaborate on the return data from the **Issuer** etc.
 
**1.** The **Merchant** creates payment request data (PREQ) and invokes the payment request API with it:
```javascript
{
  "id": "7040566321",
  "payee": "Rocket Shop",
  "amount": {
    "currency": "EUR",
    "value": "435.00"
  }
}
```
Note: the JSON object above outlines the conceptual data before being translated into a compliant Payment Request call.  It seems that `payee` currently have no obvious place in the API.

**2.** The **Browser** responds with PREQ rendered in a user interpretable fashion:
![browser-display](https://user-images.githubusercontent.com/8044211/108314345-095ede00-71ba-11eb-8b37-a0fd7ed58621.png)
_Non-normative UI_.

**3.** The **User** authorizes the request with a fingerprint, PIN, or similar.

**4.** After successful user authorization, the **Browser** _MUST_ perform the following _internal steps_.

**4.1** Calculate a Base64Url-encoded SHA256 hash of PREQ, canonicalized according to [RFC8785](https://tools.ietf.org/html/rfc8785).  Expressed as a formula: BASE64URL(SHA256(RFC8785(PREQ))). Applied to the sample data, the result of this operation should be `"Zf6qUOoA3NWM77rksuji2tE7tlGHrzu9K5Y_AGn5tkw"`.

**4.2** Create an authorization data object (AD) with data needed for processing by the **Issuer**:
```javascript
{
  "requestHash": {
    "algorithm": "S256",
    "value": "Zf6qUOoA3NWM77rksuji2tE7tlGHrzu9K5Y_AGn5tkw"
  },
  "payeeHost": "spaceshop.com",
  "accountId": "FR7630002111110020050014382",
  "timeStamp": "2021-02-18T10:32:35+01:00"
}
```
- `requestHash` cryptographically binds PREQ to AD. 
- `payeeHost` is picked-up by the **Browser**.
- `accountId` denotes the actual account number to be debited and is retrieved from the payment credential.  The  sample uses a French IBAN number but it could equally well be a credit card PAN.
- `timeStamp` is generated internally.

**4.3** Using the FIDO signature key, sign AD using [JWS/CT](https://tools.ietf.org/html/draft-jordan-jws-ct-02) resulting in the following completed AD object:
```javascript
{
  "requestHash": {
    "algorithm": "S256",
    "value": "Zf6qUOoA3NWM77rksuji2tE7tlGHrzu9K5Y_AGn5tkw"
  },
  "payeeHost": "spaceshop.com",
  "accountId": "FR7630002111110020050014382",
  "timeStamp": "2021-02-18T10:32:35+01:00",
  "signature": "eyJhbGciOiJFUzI1NiIsImp3ayI6eyJrdHkiOiJFQyIsImNydiI6IlAtMjU2IiwieCI6IjZCS3hwdHk4Y0ktZXhEekNraC1nb1U2ZFhxM01iY1kwY2QxTGFBeGlOclUiLCJ5IjoibUNiY3ZVem00NGozTHQyYjVCUHlRbG9ROTF0ZjJEMlYtZ3plVXhXYVVkZyJ9fQ..XvwOEMX4CG8l5nq-rOSkEbdFRcrhtAdQ6s1HFtLgiWMoNliiPSJnurmBnkBzudoLucBLVRwql6ZWANq1AALcXA"
}
```
Although not visible in the listing above (due to JWS' Base64Url-encoding requirements on JWS header data), AD signatures _MUST_ provide a key identifier.  In the sample, the following JWK holding the public key was used:
```javascript
{
  "kty": "EC",
  "crv": "P-256",
  "x": "6BKxpty8cI-exDzCkh-goU6dXq3MbcY0cd1LaAxiNrU",
  "y": "mCbcvUzm44j3Lt2b5BPyQloQ91tf2D2V-gzeUxWaUdg"
}
```

**4.4** For privacy and security reasons, the completed AD object _MUST_ be encrypted by a public key provided by the **Issuer** through the payment credential.  This specification builds on JWE ([RFC7516](https://tools.ietf.org/html/rfc7516)) using the compact serialization mode.  The selection between an RSA or ECDH encryption scheme is at the discretion of the **Issuer**.  AD objects _MUST_ be JSON-serialized to UTF-8 before being encrypted.  The resulting JWE string would look something like this (_with line breaks for display purposes only_):
```code
eyJhbGciOiJSU0EtT0FFUCIsImVuYyI6IkEyNTZHQ00ifQ.OKOawDo13gRp2oja
HV7LFpZcgV7T6DVZKTyKOMTYUmKoTCVJRgckCL9kiMT03JGeipsEdY3mx_etLbb
WSrFr05kLzcSr4qKAq7YN7e9jwQRb23nfa6c9d-StnImGyFDbSv04uVuxIp5Zms
1gNxKKK2Da14B8S4rzVRltdYwam_lDp5XnZAYpQdb76FdIKLaVmqgfwX7XWRxv2
322i-vDxRfqNzo_tETKzpVLzfiwQyeyPGLBIO56YJ7eObdv0je8860ppamavo35
UgoRdbYaBcoh9QcfylQr66oc6vFWXRcZ_ZT2LawVCWTIy3brGPi6UklfCpIMfIj
f7iGdXKHzg.48V1_ALb6US04U3b.5eym8TW_c8SuK0ltJ3rpYIzOeDQz7TALvtu
6UG9oMo4vpzs9tX_EFShS8iB7j6jiSdiwkIr3ajwQzaBtQD_A.XFBoMYUZodetZ
dvTiFvSkQ
```
To simplify documentation, the JWE string is subsequently referred to as `"JWE-Encrypted-AD"`.

_Note that due to the end-2-end encryption scheme, the domain constraints usually associated with FIDO, do not apply_.

**4.5** Put the encrypted AD into a payment response object (PRES) together with clear text data needed for payment "backend" processing:
```javascript
{
  "encryptedAuthorization": "JWE-Encrypted-AD",
  "paymentMethod": "https://europeanpaymentsinitiative.eu/spc",
  "issuerId": "https://mybank.fr/payment"
}
```
`paymentMethod` is the actual SPC method used.  See https://github.com/w3c/webauthn/issues/1570#issuecomment-780611040.

Note that by encrypting authorization data, the "routing" of payment requests with respect to the **Issuer** is separated from account numbers.  By using URLs like in the example, the need for databases holding BIN numbers is eliminated.

**5.** The **Browser** returns PRES to the **Merchant**.

**6.** The **Merchant** puts its original PREQ and the received PRES into an authorization request object (AREQ):
```javascript
{
  "paymentRequest": {
    "id": "7040566321",
    "payee": "Rocket Shop",
    "amount": "435.00",
    "currency": "EUR"
  },
  "encryptedAuthorization": "JWE-Encrypted-AD",
  "paymentMethod": "https://europeanpaymentsinitiative.eu/spc",
  "issuerId": "https://mybank.fr/payment"
}
```

**7.**  The **Merchant** sends AREQ to the **Acquirer** service associated with the specific `paymentMethod`.  _Note that authentication of AREQ is out of scope for this specification_.

**8.** The **Acquirer** validates the authenticity and syntax of the received AREQ as well as looking up the **Issuer** through the supplied `issuerId`.

**9.** The **Acquirer** augments the received AREQ with additional data (like **Merchant** collection account), and forwards the completed object to the designated **Issuer**.  _Note that authentication of AREQ is out of scope for this specification_.

**10.** The **Issuer** _MUST_ now perform a number of steps to verify the correctness of the received AREQ. With respect to the **User** the following steps are the most significant:

**10.1** Decrypt the `encryptedAuthorization` object in AREQ using the **Issuer** specific (but shared) decryption key.  Set result as AD.

**10.2** Retrieve the signature key declared in the JWS header of the AD `signature` element.  Abort if unknown.

**10.3** Verify that the retrieved signature key is associated with the AD `accountId` element.  Abort on failure.

**10.4** Validate the AD signature using the retrieved signature key.  Abort on failure.

**10.5** Perform the same calculation as specified in **4.1** on the received `paymentRequest` object and compare the result with the AD `requestHash`.  Abort on failure.

**10.6** Check that the AD `timeStamp` element is within limits. Tentative: _CurrentTime_ - 600s < `timeStamp` < _CurrentTime_ + 60s.  Abort on failure.

**10.7** To cope with possible replays, check that the received AD has not already been utilized.  Ideally, the **Issuer** should support idempotent operation to facilitate benign transaction retries.

Available funds also needs to be validated but that is out of scope for this specification.
@equalsJeffH @nadalin @rlin1

**Comment by adrianhopebailie** [comment]
How does the browser determine which payment credential to use in step 2?
My assumption is that the user has multiple and they are only usable on specific payment networks for specific payment method.s

**Comment by cyberphone** [comment]
Right, this proposal does not describe the selection mechanism because the assumption is that the existing PaymentRequest selection scheme will be sufficient.

**Comment by cyberphone** [comment]
Although security for the "backend" is not covered here, there are a few apparent things worth mentioning.
- The name "Super Shop" could be a lie.  The **Acquirer** should have the information needed to block such a transaction.
- The domain "supershop.com" can be provided by the **Acquirer** as a part of the augmented AREQ sent to the **Issuer**.  A mismatch should abort the transaction request.
- If the **Merchant** changes anything in its payment claim, it will be revealed by the **Issuer** due to the hashing arrangement.

**Comment by cyberphone** [comment]
@adrianhopebailie I was probably a bit quick regarding the selection thing...

Could this work?
```code
const methodData = [{
  supportedMethods: "https://emvco.com/spc",
  data: { 
      supportedNetworks: ["visa", "mastercard"] 
  }
}, {
    supportedMethods: "https://europeanpaymentsinitiative.com/spc"
}]
```
Both methods would then subscribe to the SPC method during payment credential enrollment.  This would enable merchants to either use individual "pay buttons" for each payment method, or support collections of useful methods to their liking.

**Comment by equalsJeffH** [comment]
ISTM that this issue ought to more properly be filed at https://github.com/rsolomakhin/secure-payment-confirmation

**Comment by cyberphone** [comment]
@equalsJeffH it may seem like that but the fact is that mine[1] and Google's[2] take on SPC only have the _name_ in common.

As a thought experiment you could try to envision what users (UI-wise) would have to put off with and banks would have to build (giant central registries of enrolled accounts) in order to apply 3DS/step-up authentication to IBAN accounts like the sample's `FR7630002111110020050014382`.

/Anders

1] Effectively a "technology refresh" of EMV adapted for the Web.  EMV remains to date the only payment _authorization_ scheme that enjoys both real and de-facto standards status.

2] Builds on leveraging two technologies that were designed for _other_ purposes.  FIDO: User authentication rather than payment authorization.  3DS: Supporting users who do NOT have access to sophisticated client-side crypto which was a great idea back in 1998 when it was created.
@balfanz @ve7jtb @dturnerx

**Comment by rsolomakhin** [comment]
@cyberphone, if I understand the gist of your proposal, there appears to be two primary differences from [SPC](https://github.com/rsolomakhin/secure-payment-confirmation):

1. The challenge comes from the merchant instead of the bank.
1. The browser stores some extra data for each payment credential, so it can return user's account number, encrypted with the public key of the bank, too.

Is that a fair summary?

I think perhaps this proposal would be more clear if the enrollment step was described or specified. How would it work?

**Comment by cyberphone** [comment]
Hi @rsolomakhin, I believe the differences are much bigger.  
1. 3DS/SPC is an _additional_ (and _optional_) card-holder authentication step introduced _before_ the actual payment transaction: https://stripe.com/docs/payments/3d-secure.  It is possible that Stripe' implementation does something else but the current state-diagrams are pretty hard to follow so I'm not able to tell 🤔  I have not the faintest idea how SPC could deal with SEPA and similar A2A (account to account) payments.

2. EMV is a _unified_ and _optimized_ single-step, "store-and-forward" payment request requiring no prior interaction with the issuer or handing out account numbers by users.  In my take on the matter (FIDO Web Pay), the browser would together with FIDO tokens form a _universal wallet_. https://github.com/rsolomakhin/secure-payment-confirmation/issues/33.

The differences on the browser side are rather moderate since both concepts builds on FIDO tokens and a secure browser-resident payment UI.  Enrollment is yet to be described but AFAICT the major difference is in the payment handler database since SPC defines a single payment method while FIDO Web Pay makes the FIDO tokens define payment methods, all linked to the built-in payment handler.
@dturnerx

**Comment by cyberphone** [comment]
@rsolomakhin Major issues with both solutions is that:
- Most banks already have deployed mobile payment authorization systems limiting their interest in investing in new systems that "does almost the same thing".
- Most people still use "PCs" for shopping making QR a fairly necessary part of the plot.

FIDO Web Pay (in contrast to SPC), should be attractive for Open Banking TPPs who could finally get away from systems requiring users to select bank in order to make a purchase.  Note though that Open Banking is a rather slow boat...

Personally I believe the code that is currently intended to be a part of the browser, _should long-term rather be a part of the client platform_ because then the same scheme and "virtual cards" could be used everywhere.  Such a change would be invisible to all actors.

Conclusion: there is no "low-hanging fruit" or quick wins here.  However, the availability of:
- Built in FIDO token support
- Built in universal Payment Handler

should be of interest for all but suppliers of proprietary payment systems.

**Comment by rsolomakhin** [comment]
@cyberphone - What do you mean by a built-in universal payment handler? I wonder whether that make the browser into yet another payment app that websites have to support, a la https://xkcd.com/927/.

**Comment by cyberphone** [comment]
> What do you mean by a built-in universal payment handler? I wonder whether that make the browser into yet another payment app that websites have to support

@rsolomakhin It is possible that I got this entirely wrong 🙃 but my interpretation of the SPC explainer is that:
- SPC depends on a built-in (aka "hardwired") payment handler
- SPC requires specific support not only by websites, but by acquirers and banks as well

My idea is simply making the built-in code _more universal_ based on the assumption that _user authorization_ of card payments and for example SEPA credit transfers do not motivate separate handlers since they (AFAICT...) only differ on account data and named method.

Although Saturn is just a PoC, I consider the concept above fully verified.  The Saturn wallet doesn't not know anything about specific payment systems and it still works with quite different backend flows.  BTW, this is a yet another reason to why I couldn't use `PaymentRequest` for Android as intended: it only supports statically declared methods 😒 Updating manifests or apps is not an option if you want a neutral (=universal) Wallet.
@adrianhopebailie

**Comment by cyberphone** [comment]
If my interpretation of 3DS/SPC is correct, the following should give you an indication of the huge differences in complexity between EMV/FIDO and 3DS/SPC: 
![3ds](https://user-images.githubusercontent.com/8044211/109770777-45ebfa00-7bfc-11eb-83a5-febd7bcd656b.png)

@rsolomakhin @equalsJeffH

**Comment by cyberphone** [comment]
@equalsJeffH @ianbjacobs @dturnerx @rsolomakhin @nadalin @balfanz  I would rather claim that whatever FIDO solution the Web Payment WG takes on it will be _perceived_ as the "official standpoint" for the (entire of) **W3C**, **FIDO alliance**, **EMVCo**, and **Google**, which IMO motivates an extended discussion. BTW, I have now cleaned-up the "spec" above (it has numerous of flaws): https://fido-web-pay.github.io/  It might be worth knowing that it does things not even Google and Apple Pay can do  😁 As expressed on LinkedIn in a somewhat market-ish way:
![FWP](https://fido-web-pay.github.io/specification/images/fwp-li.svg)

**Comment by rsolomakhin** [comment]
@cyberphone wrote:

>  I would rather claim that whatever FIDO solution the Web Payment WG takes on it will be perceived as the "official standpoint" for the (entire of) ...

At this stage, perhaps it would be more accurate to state that the organizations are exploring the benefits of FIDO through discussions and experimentation.

**Comment by cyberphone** [comment]
Yes @rsolomakhin,  the mentioned organizations _may_ have had discussions but the results of these have not been made public.

You probably do not agree, but I consider the combination of 
- FIDO
- PaymentRequest
- Browser-resident payment handler

as W3C's _last_ (and certainly the _best_) chance moving the payment industry forward.  That is, the stakes are **sky-high!**

Although I'm just a WPWG mailing-list subscriber, I get the impression (by reading between the minutes) that everything else is pretty much on hold making SPC a likely "default".

**Comment by equalsJeffH** [comment]
On 2021-03-17 call:
The WG consensus is that these payments-oriented discussions are not relevant to the webauthn WG and belong elsewhere, e.g., https://github.com/w3c/payment-request/issues and/or https://github.com/rsolomakhin/secure-payment-confirmation/issues

**Comment by cyberphone** [comment]
@equalsJeffH @emlun @nadalin The raised issues are not related to the PaymentRequest API since nothing new is requested. Filing an issue on the Secure Payment Confirmation (SPC) project that effectively suggests _dropping the idea altogether_ would be pretty strange. This is clearly a _charter_ question involving not only the W3C, but the FIDO alliance, the Chromium team, and EMVCo as well.  

Unlike SPC, Google pay is (with respect to the user) an _end-2-end secured_ solution and it even _encrypts_ authorizations.

I wish you and the W3C luck with the project "Making 3D Secure great again" 🤣

Leaving the topics here
/Anders

**Comment by cyberphone** [comment]
The requirements imposed by SPC makes SPC an integral part of a revised WebAuthn charter, otherwise SPC simply put won't work.  Some of the proposals aired, even affect FIDO/CTAP.

Before you include SPC in the WebAuthn charter, I would suggest taking a peek at https://github.com/mozilla/standards-positions/issues/570#issuecomment-972578433 which summarizes the somewhat voluminous data above, into a short table, which also includes a comparison with Apple Pay.
@YubicoDemo @nadalin @ve7jtb @selfissued 

To date unresolved SPC issues include:
- Card number privacy: https://github.com/w3ctag/design-reviews/issues/675#issuecomment-964273692
- Support for account-to-account payments: https://github.com/w3c/secure-payment-confirmation/issues/166
